hydra/test_utils/test_utils.py

```diff
@@ -463,3 +464,21 @@ def assert_regex_match(
                 from_name=from_name,
                 to_name=to_name,
             )
+
+
+def assert_multiline_regex_search(
+    pattern: str, string: str, from_name: str = "Expected", to_name: str = "Actual"
+) -> None:
```

Comment:
can you give more details why this new function is needed and the existing one not work for it?

Reply:
The existing function processes the inputs line-by-line, using `zip` to match each line of the regex pattern against the corresponding line in the input string.

This was a problem because different CI platforms produce warning messages with different numbers of lines. For example, the test_linux-3.9 CI job produces a two-line warning that looks like this:
```text
tests/test_apps/deprecation_warning/my_app.py:10: UserWarning: Feature FooBar is deprecated
  deprecation_warning("Feature FooBar is deprecated")
```
and the test_linux-3.8 CI job produces a one-line warning that looks like this:
```text
tests/test_apps/deprecation_warning/my_app.py:10: UserWarning: Feature FooBar is deprecated
```

This new `assert_multiline_regex_search` function gives more flexibility to match a single regex pattern against a multi-line block of text, while `assert_regex_match` fails if the regex pattern does not have the same number of lines as the input string.
In commit https://github.com/facebookresearch/hydra/pull/1703/commits/d2ea5528fb70b8ac96f02dce04d7b6e22d95327f I tried to use the existing function `assert_regex_match` instead of this new function `assert_multiline_regex_search`. For that commit, test_linux-3.9 CI job succeeded while many of the other CI jobs failed.

Reply:
I see. How does this function handle the case where the number of lines does not match?
Can it be used to replace the other implementation completely?

Reply:
> I see. How does this function handle the case where the number of lines does not match?

This function only makes one call to `re.search`. The pattern that gets matched can contain newline characters `'\n'`.
For example, I used a question-mark pattern `'\n?'` to match an "optional newline" in one of the new `test_hydra.py` test cases.

> Can it be used to replace the other implementation completely?

Yes, that should be possible.